lib/bap_disasm/bap_disasm_x86_lifter.ml

```diff
@@ -0,0 +1,2319 @@
+         * infer_ast, so I'm passing those in directly now.
+         * Imo, this is a lot less dubious than the thingy above. *)
+        let (index, index_width) = match t with
+          | Type.Imm 64 -> (Exp.Extract (((i*8)+2), ((i*8)+0), order_e), 64) (* 3 bits *)
+          | Type.Imm 128 -> (Exp.Extract (((i*8)+3), ((i*8)+0), order_e), 128) (* 4 bits *)
+          | Type.Imm 256 -> (Exp.Extract (((i*8)+3), ((i*8)+0), order_e), 256) (* 4 bits *)
+          | _ -> disfailwith "invalid size for pshufb"
+        in
+        let index = Exp.(Cast (Cast.UNSIGNED, !!t, index)) in
+        let atindex = extract_byte_symbolic_with_width dst_e index index_width in
+        Exp.Ite (highbit, it 0 8, atindex)
+      in
+      let n = (bits_of_width t) / 8 in
+      let e = Util.concat_explist (List.map ~f:get_bit (List.range ~stride:(-1) ~start:`exclusive ~stop:`inclusive n 0)) in
+      (match vsrc with
+       | None -> [assn t dst e]
+       | Some vdst -> [assn t vdst e])
+    | Lea(t, r, a) when pref = [] ->
+      (* See Table 3-64 *)
+      (* previously, it checked whether addrbits > opbits before the cast_low.
+       * The conclusion we came to was that
+         - if they were equal, the cast is basically a nop
+         - if it was the other way round, you want to extend it anyway.
+       * (I may be remembering things wrongly) *)
+      [assn t r Exp.(Cast (Cast.LOW, !!t, a))]
+    | Call(o1, ra) when pref = [] ->
+      (* If o1 is an immediate, we should syntactically have Jump(imm)
+         so that the CFG algorithm knows where the jump goes.  Otherwise
+         it will point to BB_Indirect.
+
+         Otherwise, we should evaluate the operand before decrementing esp.
+         (This really only matters when esp is the base register of a memory
+         lookup. *)
+      let target = op2e mt o1 in
+      (match o1 with
+       | Oimm _ ->
+         [Stmt.Move (rsp, Exp.(rsp_e - (Int (mi (bytes_of_width mt)))));
+          store_s mode None mt rsp_e (Exp.Int ra);
+          Stmt.Jmp target]
+       | _ ->
+         let t = Var.create ~tmp:true "target" mt in
+         [Stmt.Move (t, target);
+          Stmt.Move (rsp, Exp.(rsp_e - (Int (mi (bytes_of_width mt)))));
+          store_s mode None mt rsp_e (Exp.Int ra);
+          Stmt.Jmp (Exp.Var t)])
+    | Jump(o) ->
+      [Stmt.Jmp (jump_target mode ss has_rex o)]
+    | Jcc(o, c) ->
+      [Stmt.If (c, [Stmt.Jmp (jump_target mode ss has_rex o)], [])]
+    | Setcc(t, o1, c) ->
+      [assn t o1 Exp.(Cast (Cast.UNSIGNED, !!t, c))]
+    | Shift(st, s, dst, shift) ->
+      let open Exp.Binop in
+      assert (List.mem [r8; r16; r32; r64] s);
+      let origCOUNT, origDEST = Var.create ~tmp:true "origCOUNT" s,
+                                Var.create ~tmp:true "origDEST" s in
+      let s' = bits_of_width s in
+      let size = it s' s' in
+      let s_f = Exp.(match st with LSHIFT -> (lsl)  | RSHIFT -> (lsr)
+                                 | ARSHIFT -> (asr) | 
```

Comment:
Consider using functional constructor whenever possible, i.e., prefer `Stmt.move` to `Stmt.Move`